### PR TITLE
Migrate to postgres bitstring perms

### PR DESCRIPTION
**Pull request by l4l**
### Description of the Change

At WSV there've been used enum types similar to SM ones. It's probably inefficient and error prone (code hard to read and permissions need to be maintained).
This PR introduces behavior based on postgres bitstrings: conversion is built-in and querying is simply perfromed via boolean operations.

### Benefits

Efficiency, no need in postgres types maintaining, cleaner code, a bit better interface for PermissionSet

### Possible Drawbacks

SQL queries become more complicated :(

### CODE REVIEW

**Comment by l4l** [comment]
> ![MINOR][MINOR] [permissions.cpp#L31](https://github.com/hyperledger/iroha/blob/baef0cf9b68b61678c0fb170bc97c0fae011227f/shared_model/interfaces/impl/permissions.cpp#L31): The function 'permissionOf' is never used. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cppcheck%3AunusedFunction)

That is going to be fixed in #1500

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: shared_model/interfaces/permissions.hpp [hunk 1]
```diff
@@ -98,9 +99,13 @@ namespace shared_model {
      public:
       PermissionSet();
       PermissionSet(std::initializer_list<Perm> list);
+      PermissionSet(const std::string &s);
```
Comment (nickaleks) [inline shared_model/interfaces/permissions.hpp]: Please mark it explicit and maybe rename parameter to "bitstring" for more clarity.

Also I do not know whether or not constructing permission set from bitstring is the right abstraction for this class?

File: shared_model/interfaces/permissions.hpp [hunk 2]
```diff
@@ -98,9 +99,13 @@ namespace shared_model {
...
 
-      size_t size() const;
+      std::string to_string() const;
```
Comment (nickaleks) [inline shared_model/interfaces/permissions.hpp]: Naming does not follow our codestyle. Also maybe specify that it is a bitstring?

Again maybe the wrong abstraction here

File: shared_model/interfaces/permissions.hpp (hunk 1 above)
Comment (l4l) [reply shared_model/interfaces/permissions.hpp]: The following code won't work then:
https://github.com/hyperledger/iroha/blob/baef0cf9b68b61678c0fb170bc97c0fae011227f/irohad/ametsuchi/impl/postgres_wsv_query.cpp#L99-L100

File: irohad/ametsuchi/impl/postgres_wsv_command.cpp [hunk 3]
```diff
@@ -77,34 +76,16 @@ namespace iroha {
...
-          + generate_perm_string(entry) + ";");
+          "INSERT INTO role_has_permissions(role_id, permission) VALUES ("
+          + transaction_.quote(role_id) + ", " + transaction_.quote(perm_str)
+          + " );");
 
       auto message_gen = [&] {
         return (boost::format("failed to insert role permissions, role "
                               "id: '%s', permissions: [%s]")
-                % role_id % generate_perm_string([](auto a) { return a; }))
+                % role_id % perm_str)
```
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/ametsuchi/impl/postgres_wsv_command.cpp]: yep, but there's no dependency on backend

File: irohad/ametsuchi/impl/postgres_wsv_command.cpp [hunk 4]
```diff
@@ -116,23 +97,27 @@ namespace iroha {
...
+      const auto perm_str =
+          shared_model::interface::GrantablePermissionSet({permission})
+              .to_string();
+      auto query =
+          (boost::format(
+               "INSERT INTO account_has_grantable_permissions as "
+               "has_perm(permittee_account_id, account_id, permission) VALUES "
+               "(%1%, %2%, %3%) ON CONFLICT (permittee_account_id, account_id) "
+               "DO UPDATE SET permission=(SELECT has_perm.permission | %3% "
+               "WHERE has_perm.permission & %3% <> %3%);")
```
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/ametsuchi/impl/postgres_wsv_command.cpp]: which operations look confusing?

[the remark this answers is not in the document]
Comment (nickaleks) [reply irohad/ametsuchi/impl/postgres_wsv_command.cpp]: WHERE condition

File: irohad/ametsuchi/impl/postgres_wsv_command.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (nickaleks) [reply irohad/ametsuchi/impl/postgres_wsv_command.cpp]: Can't we make it transparent? For example define class which converts to string and parametrize it with concrete implementation?

[the remark this answers is not in the document]
Comment (lebdron) [reply irohad/ametsuchi/impl/postgres_wsv_command.cpp]: @nickaleks it sounds like another subtask for dependency injection story to me

[the remark this answers is not in the document]
Comment (lebdron) [reply irohad/ametsuchi/impl/postgres_wsv_command.cpp]: As a "temporary" solution we can use existing proto::toString method not to confuse the clients, and add TODO for more clear solution in the future

File: shared_model/interfaces/permissions.hpp [hunk 5]
```diff
@@ -96,9 +97,13 @@ namespace shared_model {
...
       PermissionSet();
       PermissionSet(std::initializer_list<Perm> list);
+      explicit PermissionSet(const std::string &bitstring);
 
-      size_t size() const;
+      std::string toBitstring() const;
+
+      static size_t size();
       PermissionSet &reset();
+      PermissionSet &set();
```
Comment (nickaleks) [inline shared_model/interfaces/permissions.hpp]: Please add comments to methods

Comment (l4l) [reply shared_model/interfaces/permissions.hpp]: I better do that in following pr, if you don't mind

Comment (nickaleks) [reply shared_model/interfaces/permissions.hpp]: Sure

File: shared_model/interfaces/permissions.hpp [hunk 6]
```diff
@@ -9,6 +9,7 @@
 #include <bitset>
 #include <functional>
 #include <initializer_list>
+#include <string>
 #include <vector>
```
Comment (nickaleks) [inline shared_model/interfaces/permissions.hpp]: Unused Include

File: irohad/ametsuchi/impl/postgres_wsv_query.cpp [hunk 7]
```diff
@@ -54,15 +53,18 @@ namespace iroha {
         const AccountIdType &permitee_account_id,
         const AccountIdType &account_id,
         shared_model::interface::permissions::Grantable permission) {
+      const auto perm_str =
+          shared_model::interface::GrantablePermissionSet({permission})
+              .toBitstring();
+      static const auto kEmpty =
```
Comment (lebdron) [inline irohad/ametsuchi/impl/postgres_wsv_query.cpp]: It is unused, so it can be removed.

**Pull request merged**